README.md
# tleft

Super naive CLI aplication that subtracts timestamps with the format `hh:mm`.

## Usage

Compile it with `stack build` and run `tleft time1 time2...`. A message with the format `Time left -> <hh>:<mm> hours` will be displayed on the screen.

## Example

```CMD
$ tleft 20:12 17:15
Time left: 02:57 hours
```


ChangeLog.md
# Changelog for tleft

## Unreleased changes


test/Spec.hs
import Test.Hspec        (Spec, it, shouldBe)
import Test.Hspec.Runner (configFastFail, defaultConfig, hspecWith)

import Tleft (tleft)

main :: IO ()
main = hspecWith defaultConfig {configFastFail = True} specs

specs :: Spec
specs = do
     it "Substrac two times" $
      tleft ["10:10", "09:02"] `shouldBe` Just (1, 8)

     it "Substrac two times lesser hour greater minute" $
      tleft ["10:10", "09:11"] `shouldBe` Just (0, 59)

     it "Substract times stamps using one digit hours" $
      tleft ["10:10", "9:02"] `shouldBe` Just (1, 8)

     it "Substract times stamps using one digit minut" $
      tleft ["10:1", "9:2"] `shouldBe` Nothing 

     it "Substrac same time" $
      tleft ["10:10", "10:10"] `shouldBe` Just (0, 0)

     it "First time bigger than the second" $
      tleft ["10:10", "11:10"] `shouldBe` Just (-1, 0)

     it "Substrac three times" $
      tleft ["10:10", "09:02", "01:00"] `shouldBe` Just (0, 8)

     it "Pass singleton list" $
      tleft ["10:10"] `shouldBe` Just (10, 10) 

     it "Pass words" $
      tleft ["he:llo"] `shouldBe` Nothing 

src/Lib.hs
module Lib
    ( tleft
    ) where

import Text.Read ( readMaybe )
import Data.Char ( isNumber )

tleft :: [String] -> Maybe (Int, Int)
tleft args
    | areTimestamps args = Just (foldl1 timeLeft (map fromArgToTimeTuple args))
    | otherwise = Nothing 

areTimestamps :: [String] -> Bool
areTimestamps [x] = isValidTime x
areTimestamps (x:xs) = isValidTime x && areTimestamps xs

isValidTime :: String -> Bool
isValidTime [h, h', colon, m, m']
    | isValidTime' [h, h'] [colon] [m, m'] = True
    | otherwise = False
isValidTime [h, colon, m,m']
    | isValidTime' [h] [colon] [m, m'] = True
    | isValidTime' [h] [colon] [m] = True
    | otherwise = False
isValidTime str = False

isValidTime' :: String -> String -> String -> Bool
isValidTime' h colon m = isValidHour h && colon == ":" && isValidMinute m

isValidHour :: String -> Bool
isValidHour a
    | isInt a = castedString >= 0
    | otherwise = False 
    where castedString = read a :: Int

isValidMinute :: String -> Bool
isValidMinute a
    | isInt a = castedString >= 0  && castedString <= 59 
    | otherwise = False 
    where castedString = read a :: Int

isInt :: String -> Bool
isInt = foldl (\acc c -> isNumber c) False

fromArgToTimeTuple :: [Char] -> (Int, Int)
fromArgToTimeTuple arg = (fromArgToHours arg, fromArgToMinutes arg)

fromArgToHours :: [Char] -> Int
fromArgToHours [x, _, _, _] = read [x]
fromArgToHours [x, y, _, __, _] = read [x, y]

fromArgToMinutes :: [Char] -> Int
fromArgToMinutes [_, _, _, m, m'] = read [m, m']
fromArgToMinutes [_, _, m, m'] = read [m, m']

timeLeft :: (Int, Int) -> (Int, Int) -> (Int, Int)
timeLeft time1 time2 = (hours, minutes) where 
    minutes = let minutes' = subMinutes (snd time1) (snd time2) in
        if minutes' < 60 then minutes' else 00
    hours = if snd time1 >= snd time2 
            then fst time1 - fst time2 
            else fst time1 - fst time2 - 1

subMinutes :: Int -> Int -> Int 
subMinutes m1 m2 = if m1 > m2 then m1 - m2 else m1 - m2 + 60

hoursLeft :: Int -> Int -> Int 
hoursLeft bigerHour smallerHour = bigerHour - smallerHour

app/Main.hs
module Main where

import System.Environment ( getArgs )
import Data.Maybe ( fromJust, isJust )
import Tleft (tleft)

main :: IO ()
main = do
    args <- getArgs 
    argParser args

usage :: [Char]
usage = "Usage: \n\t tleft <hh:mm> <hh:mm>"

argParser :: [String] -> IO ()
argParser [x]
    | x `elem` helpCommands = putStrLn usage
    | otherwise = putStrLn ("Error: Must specify two or more timestamps.\n" ++ usage)
    where helpCommands = ["--help", "-h"]
argParser args@(x:xs)
    | isJust time = putStrLn (composeTimeLeftString (fromJust time))
    | otherwise = putStrLn ("Error: Expecting two timestamps. Please check usage:\n" ++ usage)
    where time = tleft args

composeTimeLeftString :: (Int, Int) -> String
composeTimeLeftString time = "Time left -> " ++ twoDigitFormatt (fst time) ++ ":" ++ twoDigitFormatt (snd time) ++ " hours"

twoDigitFormatt :: Int -> String
twoDigitFormatt n
    | n >= 10 || n < 0 = show n
    | otherwise = "0" ++ show n

src/Tleft.hs
module Tleft
    ( tleft
    ) where

import Text.Read ( readMaybe )
import Data.Char ( isNumber )

tleft :: [String] -> Maybe (Int, Int)
tleft args
    | areTimestamps args = Just . foldl1 timeLeft $ map fromArgToTimeTuple args
    | otherwise = Nothing 

areTimestamps :: [String] -> Bool
areTimestamps [x] = isValidTime x
areTimestamps (x:xs) = isValidTime x && areTimestamps xs

isValidTime :: String -> Bool
isValidTime [h, h', colon, m, m']
    | isValidTime' [h, h'] [colon] [m, m'] = True
    | otherwise = False
isValidTime [h, colon, m,m']
    | isValidTime' [h] [colon] [m, m'] = True
    | isValidTime' [h] [colon] [m] = True
    | otherwise = False
isValidTime str = False

isValidTime' :: String -> String -> String -> Bool
isValidTime' h colon m = isValidHour h && colon == ":" && isValidMinute m

isValidHour :: String -> Bool
isValidHour a
    | isInt a = castedString >= 0
    | otherwise = False 
    where castedString = read a :: Int

isValidMinute :: String -> Bool
isValidMinute a
    | isInt a = castedString >= 0  && castedString <= 59 
    | otherwise = False 
    where castedString = read a :: Int

isInt :: String -> Bool
isInt = foldl (\acc c -> isNumber c) False

fromArgToTimeTuple :: [Char] -> (Int, Int)
fromArgToTimeTuple arg = (fromArgToHours arg, fromArgToMinutes arg)

fromArgToHours :: [Char] -> Int
fromArgToHours [x, _, _, _] = read [x]
fromArgToHours [x, y, _, __, _] = read [x, y]

fromArgToMinutes :: [Char] -> Int
fromArgToMinutes [_, _, _, m, m'] = read [m, m']
fromArgToMinutes [_, _, m, m'] = read [m, m']

timeLeft :: (Int, Int) -> (Int, Int) -> (Int, Int)
timeLeft time1 time2 = (hours, minutes) where 
    minutes = let minutes' = subMinutes (snd time1) (snd time2) in
        if minutes' < 60 then minutes' else 00
    hours = if snd time1 >= snd time2 
            then fst time1 - fst time2 
            else fst time1 - fst time2 - 1

subMinutes :: Int -> Int -> Int 
subMinutes m1 m2 = if m1 > m2 then m1 - m2 else m1 - m2 + 60

hoursLeft :: Int -> Int -> Int 
hoursLeft bigerHour smallerHour = bigerHour - smallerHour

tleft.hs
{-|
Module      : tleft
Description : Super naive CLI aplication that subtracts two timestamps with the format hh:mm.
Copyright   : (c) Felipe-Cruz Martínez-Acalá, 2021
License     : MIT license
Maintainer  : fesanmar@gmail.com
Stability   : experimental
Version     : 1.0.1
-}

import System.Environment
import Text.Read
    
main :: IO ()
main = do
    args <- getArgs 
    argParser args

usage = "Usage: \n\t tleft <hh:mm> <hh:mm>"

argParser :: [String] -> IO ()
argParser (x:[])
    | x `elem` helpCommands = putStrLn usage
    | otherwise = putStrLn ("Error: Must specify two timestamps.\n" ++ usage)
    where helpCommands = ["--help", "-h"]
argParser (x:y:[])
    | isValidTimeRange x y = putStrLn (timeLeftFromArgs x y)
    | areTimestamps [x, y] = putStrLn ("Error: first arg <" ++ x ++ ">" ++ " must be greater or equal to the second <" ++ y ++ ">\n" ++ usage)
    | otherwise = putStrLn ("Error: Expecting two timestamps. Please check usage:\n" ++ usage)
argParser args@(x:y:zs) = putStrLn ("Error: Expecting only two args but " ++ (show (length args)) ++ " where passsed.\n" ++ usage)
 
isValidTimeRange :: String -> String -> Bool
isValidTimeRange a b = areTimestamps [a, b] && a `isBiggerThanOrEqualTo` b

isBiggerThanOrEqualTo :: String -> String -> Bool
isBiggerThanOrEqualTo a b
    | h > h' = True
    | h == h' = m > m' || m == m'
    | otherwise = False
    where 
        h = fromArgToHours a
        h' = fromArgToHours b
        m = fromArgToMinutes a
        m' = fromArgToMinutes b

areTimestamps :: [String] -> Bool
areTimestamps (x:[]) = isValidTime x
areTimestamps (x:xs) = isValidTime x && areTimestamps xs

isValidTime :: String -> Bool
isValidTime (h:h':colon:m:m':[])
    | isValidTime' [h, h'] [colon] [m, m'] = True
    | otherwise = False
isValidTime (h:colon:m:m':[])
    | isValidTime' [h] [colon] [m, m'] = True
    | isValidTime' [h] [colon] [m] = True
    | otherwise = False
isValidTime (h:colon:m:m':_) = False

isValidTime' :: String -> String -> String -> Bool
isValidTime' h colon m = isValidHour h && colon == ":" && isValidMinute m

isValidHour :: String -> Bool
isValidHour a
    | isInt a = castedString >= 0
    | otherwise = False 
    where castedString = (read a :: Int)

isValidMinute :: String -> Bool
isValidMinute a
    | isInt a = castedString >= 0  && castedString <= 59 
    | otherwise = False 
    where castedString = (read a :: Int)

isInt :: String -> Bool
isInt a = (readMaybe a :: Maybe Int) /= Nothing

timeLeftFromArgs :: [Char] -> [Char] -> [Char]
timeLeftFromArgs arg1 arg2 = timeLeftMessage (fromArgToTimeTuple arg1) (fromArgToTimeTuple arg2)

fromArgToTimeTuple :: [Char] -> (Int, Int)
fromArgToTimeTuple arg = (fromArgToHours arg, fromArgToMinutes arg)

fromArgToHours :: [Char] -> Int
fromArgToHours (x:_:_:_:[]) = read [x]
fromArgToHours (x:y:_:__:_:[]) = read [x, y]

fromArgToMinutes :: [Char] -> Int
fromArgToMinutes (_:_:_:m:m':[]) = read ([m, m'])
fromArgToMinutes (_:_:m:m':[]) = read ([m, m'])

timeLeftMessage :: (Int, Int) -> (Int, Int) -> [Char]
timeLeftMessage time1 time2 = composeTimeLeftString (timeLeft time1 time2)

composeTimeLeftString :: (Int, Int) -> [Char]
composeTimeLeftString time = "Time left -> " ++ twoDigitFormatt (fst time) ++ ":" ++ twoDigitFormatt (snd time) ++ " hours"

twoDigitFormatt :: Int -> String
twoDigitFormatt n
    | n >= 10 || n < 0 = show n
    | otherwise = "0" ++ (show n)

timeLeft :: (Int, Int) -> (Int, Int) -> (Int, Int)
timeLeft time1 time2 = (hours, minutes) where 
    minutes = let minutes' = subMinutes (snd time1) (snd time2) in
        if minutes' < 60 then minutes' else 00
    hours = if snd time1 >= snd time2 
            then fst time1 - fst time2 
            else fst time1 - fst time2 - 1

subMinutes :: Int -> Int -> Int 
subMinutes m1 m2 = if m1 > m2 then m1 - m2 else m1 - m2 + 60

hoursLeft :: Int -> Int -> Int 
hoursLeft bigerHour smallerHour = bigerHour - smallerHour
